python/lsst/ts/ess/csc/data_client/siglent_ssa3000x_spectrum_analyzer_data_client.py

```diff
@@ -177,39 +199,98 @@ async def read_loop(self) -> None:
-                    await self.write(self.query_trace_data_cmd)
-                    assert self.client is not None  # make mypy happy
-                    read_bytes = await asyncio.wait_for(
-                        self.client.readuntil(TERMINATOR),
-                        timeout=self.config.read_timeout,
+                await self.write(self.query_trace_data_cmd)
+                assert self.client is not None  # make mypy happy
+                read_bytes = await asyncio.wait_for(
+                    self.client.readuntil(TERMINATOR),
+                    timeout=self.config.read_timeout,
+                )
+                raw_data = read_bytes.decode().strip()
+                raw_data_items = raw_data.split(",")
+                data = [float(i.strip()) for i in raw_data_items]
+                if (
+                    len(data) < EXPECTED_NUMBER_OF_DATA_POINTS
+                    and self.number_of_truncated_data_read < MAX_NUM_TRUNCATED_DATA
+                ):
+                    logging.warning(
+                        f"Data of length {len(data)} read. Ignoring because this is "
+                        f"read #{self.number_of_truncated_data_read} out of {MAX_NUM_TRUNCATED_DATA}."
                     )
-                    raw_data = read_bytes.decode().strip()
-                    raw_data_items = raw_data.split(",")
-                    data = [float(i.strip()) for i in raw_data_items]
-                else:
-                    # Generate random data between 0 and -100 dB. Convert to
-                    # list to keep mypy happy.
-                    data = (-100.0 * np.random.random(self.num_data_points)).tolist()
-                try:
-                    await self.topics.tel_spectrumAnalyzer.set_write(
-                        startFrequency=self.start_frequency,
-                        stopFrequency=self.stop_frequency,
-                        spectrum=data,
-                        timestamp=timestamp,
+                    self.number_of_truncated_data_read += 1
+                elif len(data) != EXPECTED_NUMBER_OF_DATA_POINTS:
+                    raise RuntimeError(
+                        f"Encountered {len(data)} data points instead of "
+                        f"{EXPECTED_NUMBER_OF_DATA_POINTS}. Check the Spectrum "
+                        f"Analyzer and the configuration."
                     )
-                except Exception as e:
-                    self.log.exception(f"Failed to handle {data=}: {e!r}")
+                else:
+                    try:
+                        await self.topics.tel_spectrumAnalyzer.set_write(
+                            startFrequency=self.start_frequency,
+                            stopFrequency=self.stop_frequency,
+                            spectrum=data,
+                            timestamp=timestamp,
+                        )
+                    except Exception as e:
+                        self.log.exception(f"Failed to handle {data=}: {e!r}")
```

Comment:
Somewhere in here, preferably right after a read that does not time out, please reset self.num_consecutive_read_timeouts to 0 (because the streak of consecutive timeouts has been broken).

Same for the weather data client.